### PR TITLE
test(e2e): Add self-hosted e2e test

### PR DESCRIPTION
**Pull request by jimmidyson**
Creating as draft until kubernetes-sigs/cluster-api-addon-provider-helm#189 is merged and released.

Depends on #494.

### CODE REVIEW

**Review by dkoshkin — APPROVED** [review]
Thanks!

File: test/e2e/clusterautoscaler_helpers.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		client.ObjectKey{
+			Namespace: input.WorkloadCluster.Namespace,
+			Name:      input.WorkloadCluster.Name + "-kubeconfig",
+		},
+		&corev1.Secret{},
+	)
+	if err != nil && errors.IsNotFound(err) {
+		return
+	}
+	Expect(err).NotTo(HaveOccurred())
```
Comment (dkoshkin) [inline test/e2e/clusterautoscaler_helpers.go:61]: nit: we can reuse the same function that the CA addon uses
```suggestion
	// Only check for ClusterAutoscaler if the cluster is self-managed.
	// ManagementCluster function will return a nil managementCluster if workloadClusterClient
	// is not a self-managed cluster.
	workloadClusterClient := input.ClusterProxy.GetWorkloadCluster(
		ctx, input.WorkloadCluster.Namespace, input.WorkloadCluster.Name,
	).GetClient()
	managementCluster, err := utils.ManagementCluster(ctx, workloadClusterClient)
	Expect(err).NotTo(HaveOccurred())
	if  managementCluster == nil {
		return
	}
```

File: .github/workflows/checks.yml [hunk 2]
```diff
@@ -54,15 +54,35 @@ jobs:
...
           test-results: test.json
 
-  e2e-test:
+  e2e-quick-start:
     strategy:
       matrix:
         provider:
           - Docker
+          # Uncomment below once we have the ability to run e2e tests on AWS from GHA.
+          # - AWS
```
Comment (deepakm-ntnx) [inline .github/workflows/checks.yml:63]: or Nutanix. was wondering how are we planning to pass the setup details of either aws or nutanix

Comment (dkoshkin) [reply .github/workflows/checks.yml:63]: GHA Secrets should be enough I think.

File: test/e2e/clusterautoscaler_helpers.go (hunk 1 above)
Comment (jimmidyson) [reply test/e2e/clusterautoscaler_helpers.go:61]: Thanks! Applied.

File: .github/workflows/checks.yml (hunk 2 above)
Comment (jimmidyson) [reply .github/workflows/checks.yml:63]: Yes these will need to be environment variables specfified either via GHA secrets when running in GHA or just normal env vars (via a local envrc perhaps) for local development. These should be looked up via the e2e variables though as a consistent way to interact with variables either from e2e config file or from the environment.

**Comment by jimmidyson** [comment]
To merge this, will need to update required status checks (make `e2e-test (Docker)` non-required and make the `e2e-quick-start` and `e2e-self-hosted` checks required) via Canaveral, merge it, and then all other open PRs will need to rebased to include this self-managed e2e.

File: test/e2e/framework/self_hosted.go [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+			framework.CreateNamespaceAndWatchEventsInput{
+				Creator:             selfHostedClusterProxy.GetClient(),
+				ClientSet:           selfHostedClusterProxy.GetClientSet(),
+				Name:                namespace.Name,
+				LogFolder:           filepath.Join(input.ArtifactFolder, "clusters", "bootstrap"),
+				IgnoreAlreadyExists: true,
+			},
+		)
+
+		By("Loading CAPI runtime extensions image to the workload cluster")
```
Comment (faiq) [inline test/e2e/framework/self_hosted.go:230]: Is this a mistake? Where is this happening?

Comment (jimmidyson) [reply test/e2e/framework/self_hosted.go:230]: Oops yeah, this is actually happening on lines 201

**Review by deepakm-ntnx — APPROVED** [review]

**Review by dkoshkin — APPROVED** [review]

**Review by mhrabovcin — APPROVED** [review]

**Pull request merged**